bot/exts/filtering/_filter_context.py

```diff
@@ -0,0 +1,62 @@
+from __future__ import annotations
+
+import typing
+from collections.abc import Callable, Coroutine, Iterable
+from dataclasses import dataclass, field, replace
+from enum import Enum, auto
+
+from discord import DMChannel, Member, Message, TextChannel, Thread, User
+
+from bot.utils.message_cache import MessageCache
+
+if typing.TYPE_CHECKING:
+    from bot.exts.filtering._filters.filter import Filter
+
+
+class Event(Enum):
+    """Types of events that can trigger filtering. Note this does not have to align with gateway event types."""
+
+    MESSAGE = auto()
+    MESSAGE_EDIT = auto()
+    NICKNAME = auto()
+
+
+@dataclass
+class FilterContext:
+    """A dataclass containing the information that should be filtered, and output information of the filtering."""
+
+    # Input context
+    event: Event  # The type of event
+    author: User | Member | None  # Who triggered the event
+    channel: TextChannel | Thread | DMChannel | None  # The channel involved
+    content: str | Iterable  # What actually needs filtering
+    message: Message | None  # The message involved
+    embeds: list = field(default_factory=list)  # Any embeds involved
```

Comment:
Why are some of the generics here (like `list`) not parameterized? This is a chonky object, I'd at least want to know what's in it

Reply:
2bd9d23

Reply:
What about these
```py
    content: str | Iterable  # What actually needs filtering
    action_descriptions: list = field(default_factory=list)  # What actions were taken
```

Reply:
I somehow missed `action_descriptions`, but for `content` the type of Iterable depends on the filter list. It might be more useful here to expand the comment.

Reply:
659320e